### DISCUSSION TITLE
Writing trial metadata to CSV

### DISCUSSION THREAD

**Discussion by cosmouse** [bonsai-rx/bonsai#1314] [Q&A] [answered]
Hello,

I am trying to write metadata to a CSV file where each row is an individual trial. The challenge is that some of the data are changing trial-to-trial and others are not. I am using Zip to combine all of the nodes with data that change on each trial (so that it just produces one output once each node has updated its value), then am using CombineLatest to combine those data with the other metadata that do not change (animal name, etc.). The problem is that I end up with nested data going into the CSV writer, and so some of the cells have parentheses in the data where I don't want them (see attached screenshots of workflow and CSV file. Is there some way to flatten the output of CombineLatest, or get rid of the nested parentheses, to avoid this? Or a more clever way to combine the data?

Also, I haven't been able to figure out how to include a header for each column. I can set IncludeHeader to 1 but where would I feed in the header names? Is there an property I can externalize to do this?

Thanks!
Phil

![Metadata](https://user-images.githubusercontent.com/21347116/228961548-c676711c-d1bb-435f-b9da-097cc5d83dca.PNG)
![Workflow](https://user-images.githubusercontent.com/21347116/228961549-14b60b7c-86d4-4650-a634-effe1de20505.png)

**Comment by bruno-f-cruz** [comment, marked as answer]
You can use `MemberSelector` to unfold the nested structure.

`CsvWriter`uses the names of the class fields to name the columns. We can leverage this behavior, and create a `DynamicClass` before the the writer:

```
new(
it.Item1 as Column1,
it.Item2.Item1 as Column2,
it.Item3.Centroid.X as X)
```

Hope it helps,
Bruno

**Reply by cosmouse** [reply]
Thanks Bruno, the MemberSelector method worked well. Could you please clarify on the DynamicClass? I didn't see any nodes with that name and can't find anything on the website about it - does this require a C# scripting node or something?

Thanks,
Phil

**Reply by bruno-f-cruz** [reply]
You should use an expressiontransform

**Reply by cosmouse** [reply]
thanks Bruno!
